docs/specs/lcb-signals-and-events.md

```diff
@@ -0,0 +1,222 @@
+
+1. Signal emission and handling is not very well checked.  At a language level, there is no way for widgets to describe what type of signals they can emit.  Similarly, there is no syntax for distinguishing between widget handlers and "normal" handlers, meaning that (in principal) a handler that is not intended to be a signal handler could end up getting called in a signal context if a child widget happens to start emitting a signal with a matching name.  Finally, there is no facility for checking that signal prototypes match signal handler prototypes.
+
+2. Similar issues to 1. above apply for event handling.
+
+3. In LCS, event handling propagates upward: if a button doesn't have an `MouseUp` handler, then the event gets passed on to a parent card, and so on.  Furthermore, there is the idea of explicit propagation: it is possible to do something in an event handler and then, for example, `pass MouseUp` to invoke the normal event propagation logic.
+
+   In LCB, event propagation is *not* currently automatic.  If a widget does not explicitly handle and re-`post` each possible event, then its script (or parent widget) never sees the event.
+
+4. Sometimes, it's important to be able to be able to prevent child widgets from seeing an event, or performing some pre-processing / post-processing around the normal event handling process.  The current LCB event handling scheme does not permit this.
+
+## Assumptions
+
+For the purposes of this document we assume that:
+
+* All connection of signal and event handlers is implicit, rather than explicit.
+
+* An instance of a widget as an engine control can be considered as a composed widget, comprising a script and the "real" widget instance (this makes sense when you consider the fact that the widget control exists, has a script, and can be manipulated even if the "real" widget was unable to be loaded for some reason).  This will be referred to as a "host widget" in the rest of this document.
+
+## Signals
+
+### Signal emission
+
+A widget's LCB source code must explicitly name, and provide prototypes for, the signals that it can emit [1]:
+
+    public signal <name:id>(<params:formals>)
+
+Signal declarations must be placed in module scope.  The signal **name** must be unique in the module.
+
+Signals are explicitly emitted by a widget using an `emit` statement:
+
+    emit <name:id>(<params:expressions>)
+
+It is a run-time error to emit a signal with the wrong number or types of emission parameters [2].  The compiler may statically verify that signals are emitted with compatible parameter lists.
+
+The `emit` statement may return before all the signal handlers for the signal have finished running [3].
+
+The signal handlers for a signal get run exactly once for each emission.  All of the handlers for an emission must complete before any of the handlers for a subsequent emission begin [4].
+
+The signal handlers for a signal may be run in any order.
+
```

Comment:
The issue I see with an asynchronous call is that you can ensure that the data will be provided before the widget needs to display the data on the screen. If the widget is being fed from a local data source then that sort of visual delay should never happen.

But perhaps the old way of feeding data into an LCS control isn't the way to handle it with widgets.  Maybe using signals for getting data into a widget is the wrong approach. 

I've been toying with the idea of a DataProvider module that would provide a common interface that could feed data from different sources into a widget. The widget would call defined methods in the DataProvider when it wants data and the DataProvider returns it. I don't know if this helps the situation at all but I was thinking about something like this:

```
local theData

# Create a data provider from text. Assumed to be line delimited.
put "Line 1" & cr & "Line 2" & cr & "Line 3" into theData
create dataprovider "MyLineProvider" with theData

# Create a data provider from a numerically indexed array.
put "Line 1" into theData[1]["label"]
put "Line 2" into theData[2]["label"]
put "Line 3" into theData[3]["label"]
create dataprovider "MyArrayProvider" with theData

# Create a data provider that uses a callback to get data.
create dataprovider "MyDatabaseProvider" with callback "GetDataForDatabaseRow" of stack "MyDatabaseStack"
set the number of rows of dataprovider "MyDatabaseProvider" to 300

set the dataprovider of widget "MyRowView" to "MyLineProvider" | "MyArrayProvider" | "MyDatabaseProvider"
```
If a DataProvider was using text or array data then it would just return the data assigned to it. The idea of the callback is that the DataProvider would ask LCS for the data in a synchronous way. But there could also be an asynchronous way of a DataProvider getting data.

I don't know if this approach would make any more sense, but we do need to figure out a way of solving the problem if we can't get external data into a widget while the screen is locked during a scrolling event.